### PR TITLE
Add infrastructure applications to DuperModel

### PR DESCRIPTION
**Pull request by hakonhall**
DuperModel is (will be) responsible for both active tenant applications
(through SuperModel) and infrastructure applications. This PR is one step
in that direction:
 - All infrastructure applications (config, confighost, controller,
   controllerhost, and proxyhost) are owned and managed by DuperModel.
 - The InfrastructureProvisioner retrieves all possible infra apps from the
   DuperModel (through a reduced API), and "activates" each of them if
   target is set and there are any nodes etc.
 - The InfrastructureProvisioner then notifies the DuperModel which
   apps have been activated, and with which hosts.
 - The DuperModel can then build delegate artificially create ApplicationInfo,
   which gets translated into the application model, and finally the service
   model.
 - The resulting service model has NOT_CHECKED for each hostadmin service
   instance. This is sufficient for goal 1 of this sprint.
 - The config server application currently has health, so that's kept as-is
   for now.
 - Feature flags have been tried and works and allows 1. to disable adding the
   infra apps in the DuperModel, and 2. to enable the infra configserver
   instead of the currently created configserver w/health.

### CODE REVIEW

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java [hunk 1]
```diff
@@ -34,76 +31,71 @@
...
+        for (InfraApplicationApi application: duperModel.getSupportedInfraApplications()) {
             try (Mutex lock = nodeRepository().lock(application.getApplicationId())) {
-                Optional<Version> version = getTargetVersion(application.getCapacity().type());
-                if (! version.isPresent()) continue;
+                NodeType nodeType = application.getCapacity().type();
+
+                Optional<Version> targetVersion = infrastructureVersions.getTargetVersionFor(nodeType);
+                if (!targetVersion.isPresent()) {
+                    logger.log(LogLevel.DEBUG, "Skipping provision of " + nodeType + ": No target version set");
+                    duperModel.infraApplicationRemoved(application.getApplicationId());
```
Comment (freva) [inline node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: It looks like its not possible to unset a target version once it has been set, so in that regard this is fine... 

However, if that were to change, this should also actually deallocate nodes allocated to this application? As this stand right now, those nodes would still be in `active`, but `DuperModel` wouldn't know anything about them?

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisionerTest.java [hunk 2]
```diff
@@ -38,90 +53,170 @@
...
+
+        infrastructureProvisioner.maintain();
+
+        verify(provisioner).prepare(eq(application.getApplicationId()), any(), any(), anyInt(), any());
+        verify(provisioner).activate(any(), eq(application.getApplicationId()), any());
+        verify(duperModelInfraApi).infraApplicationActivated(application.getApplicationId(), Arrays.asList(node2, node3));
     }
 
     @Test
     public void returns_empty_if_no_usable_nodes() {
```
Comment (freva) [inline node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisionerTest.java]: Update name of the test

File: service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        this.useConfigserverConfig = useConfigserverConfig;
+
+        // Single-tenant applications have the config server as part of the application model, conflicting and would have conflicting.
+        // TODO: Add health monitoring for config server when part of application model.
+        if (useConfigserverConfig && multitenant) {
+            staticInfos.add(configServerApplicationInfo);
+        }
+    }
+
+    public ConfigServerApplication getConfigServerApplication() {
```
Comment (freva) [inline service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java]: Seems strange to have these methods, they are only used in tests and can be easily created by the test anyway?

File: service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+    /** For testing */
+    public DuperModel(boolean containsInfra,
+                      boolean useConfigserverConfig,
+                      boolean multitenant,
+                      ApplicationInfo configServerApplicationInfo) {
+        this.containsInfra = containsInfra;
+        this.useConfigserverConfig = useConfigserverConfig;
+
+        // Single-tenant applications have the config server as part of the application model, conflicting and would have conflicting.
```
Comment (freva) [inline service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java]: What does 

> conflicting and would have conflicting

mean?

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java (hunk 1 above)
Comment (hakonhall) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Yes, I think the InfrastructureProvisioner should then deallocate those nodes, and it sounds right that the DuperModel would not know about that application then.

File: node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisionerTest.java (hunk 2 above)
Comment (hakonhall) [reply node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisionerTest.java]: done

File: node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java (hunk 1 above)
Comment (freva) [reply node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/InfrastructureProvisioner.java]: Right, it should, but it doesn't. Remove the removal and add a TODO instead?

File: service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java (hunk 3 above)
Comment (hakonhall) [reply service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java]: If clients of DuperModel needed to get hold of e.g. the ConfigServerApplication, it seems correct to ask the DuperModel for it. There used to be public constants that exposed these applications, but they have been replaced with getters on DuperModel.

These are used in 2 tests. You mean the tests could 'new' the applications, even though the DuperModel already contains them?

File: service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java (hunk 4 above)
Comment (hakonhall) [reply service-monitor/src/main/java/com/yahoo/vespa/service/monitor/application/DuperModel.java]: Fixed: something strange happened when I copied this line: this part was added. Removed.

**Review by freva — APPROVED** [review]

**Pull request merged**